[RLlib] EnvRunnerV2 and EpisodeV2 that support Connectors.

gjoliver opened the pull request:
## Why are these changes needed?

Rewrite EnvRunner and Episode to make data flow clearer.

After the refactoring, we have:
```
Sample -> EnvRunnerV2 -> List[EpisodeV2] -> List[AgentCollector]
              |
              -> List[PolicyCollector]
```
EnvRunnerV2 is disabled by default for now to keep thing safe.

## Related issue number

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [*] Unit tests
   - [*] Release tests
   - [ ] This PR is not tested :(

sven1977 commented on rllib/algorithms/callbacks.py:
```diff
@@ -169,7 +171,7 @@ def on_episode_end(
         base_env: BaseEnv,
         policies: Dict[PolicyID, Policy],
-        episode: Episode,
+        episode: Union[Episode, EpisodeV2, Exception],
```
Can you add to the dosctring why these can also be Exceptions and in which cases?

gjoliver replied:
done. oh, yeah, I should have given you a heads up.
while doing this, I noticed that we should probably call on_episode_end() callbacks with the exception if we receive one from the env as well, since otherwise, user logic won't know the episode has errored out, and will left hanging forever.

sven1977 commented on rllib/algorithms/callbacks.py:
```diff
@@ -352,7 +354,7 @@ def on_episode_end(
         base_env: BaseEnv,
         policies: Dict[PolicyID, Policy],
-        episode: Episode,
+        episode: Union[Episode, EpisodeV2, Exception],
```
same

sven1977 commented on rllib/algorithms/callbacks.py:
```diff
@@ -138,7 +140,7 @@ def on_episode_step(
         base_env: BaseEnv,
         policies: Optional[Dict[PolicyID, Policy]] = None,
-        episode: Episode,
+        episode: Union[Episode, EpisodeV2],
```
Why don't we expcet an exception here?

sven1977 commented on rllib/connectors/action/clip.py:
```diff
@@ -1,26 +1,23 @@
 
 
-@DeveloperAPI
+@ExperimentalAPI
```
Let's not use this tag anymore.
Even better: Could we start using the generic ray decorators and start phasing out the RLlib specific ones?

sven1977 commented on rllib/connectors/tests/test_action.py:
```diff
@@ -1,18 +1,15 @@
 import gym
 import numpy as np
-import unittest
+import pytest
```
??
Can we stick to the common pattern? `unittest.Testcase`?

sven1977 commented on rllib/connectors/tests/test_action.py:
```diff
@@ -119,9 +79,27 @@ def test_clip_action_connector(self):
+
 
 if __name__ == "__main__":
-    import pytest
```
??

sven1977 commented on rllib/evaluation/env_runner_v2.py:
```diff
@@ -0,0 +1,1066 @@
+        policy_collectors: collected training SampleBatchs by policy.
+
+    Returns:
+        MultiAgentBatch. SampleBatch if this is a single-agent env.
```
Actually, now that we are rewriting this whole thing, could we always return a MultiAgentBatch here?
This ambiguity has led to a lot of confusion for users working with our `training_step` methods in the past as to what to expect from a `sample()` call.
I want to get rid of the need to always call `.as_multi_agent()`.

sven1977 commented on rllib/evaluation/episode_v2.py:
```diff
@@ -0,0 +1,335 @@
+        # Total # of steps take by all agents in this env.
+        self.total_agent_steps: int = 0
+        # Dict for user to add custom metrics.
+        self.custom_metrics: Dict[str, float] = {}
+        # Temporary storage. E.g. storing data in between two custom
+        # callbacks referring to the same episode.
+        self.user_data: Dict[str, Any] = {}
+        # Dict mapping str keys to List[float] for storage of
+        # per-timestep float data throughout the episode.
+        self.hist_data: Dict[str, List[float]] = {}
```
Maybe in a follow up PR: Can we unify these 3? It's super confusing to the users as to what they have to write to in their custom callbacks. Why don't we have one per-timestep structure that they can write into (each timestep `on_episode_step`) and one per-episode structure they can use to write more generic metrics into, like the episode return?

sven1977 commented on rllib/evaluation/sampler.py:
```diff
@@ -263,23 +240,37 @@ def __init__(
-            self.render,
-        )
+        if worker.policy_config.get("enable_connectors", False):
+            self._env_runner = EnvRunnerV2(
```
Nice to have this as an object now!

sven1977 commented:
Looks great, thanks for doing this cleanup!
Just a few nits and TODOs for follow-up PRs:
* No disambiguity in returning SampleBatch or MultiAgentBatch (should always return the latter!)
* Unify structures in episode to write into as a user that overrides custom callbacks (having `hist_stats`, `user_data`, and `custom_metrics` is confusing).
* Atari metrics should not be different from regular ones.

gjoliver reviewed (commented):
address review comments.

gjoliver commented:
> * Unify structures in episode to write into as a user that overrides custom callbacks (having `hist_stats`, `user_data`, and `custom_metrics` is confusing).
> * Atari metrics should not be different from regular ones.

ok, I changed to always return MultiAgentBatch.
And added TODOs for the next 2.
I think we need to do these after Episode1 is retired though (after the V2 stuff is well tested and turned on by default).

smorad commented on rllib/evaluation/episode_v2.py:
```diff
@@ -0,0 +1,338 @@
+            worker: The RolloutWorker instance, in which this episode runs.
+        """
+        # Unique id identifying this trajectory.
+        self.episode_id: int = random.randrange(2e9)
```
Is it possible to have truly unique episode ids? Random sampling like this doesn't make it unique, in fact I think there are some other places we do this which causes crashes/issues for long-running jobs

sven1977 reviewed (approved):
Looks good to me now. Thanks for all the refactoring work and requested changes @gjoliver !
Another huge milestone toward cleaner, more intuitive APIs.
